src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml

```diff
@@ -892,6 +899,8 @@
     <Field Name="ExpressionOpt" Type="BoundExpression" Null="allow"/>
     <Field Name="IDisposableConversion" Type="Conversion" />
     <Field Name="Body" Type="BoundStatement"/>
+    <!-- A bound await, but with a placeholder instead of an awaited expression -->
+    <Field Name="AwaitOpt" Type="BoundAwaitExpression" Null="allow"/>
```

Comment:
I wonder if it is better to factor out `{getAwaiter, isCompleted, getResult}` as an `AwaitableInfo` and use it in both BoundAwait and in BoundUsing?
GetAwaitableExpressionInfo is basically returning such info (although via bunch of `out` params).

Such approach might scale better at representing implicit awaitability than carrying around await nodes on a side.

Reply:
I'm taking a look at this and will get back to you.

---
In reply to: [160565220](https://github.com/dotnet/roslyn/pull/23961#discussion_r160565220) [](ancestors = 160565220)

Reply:
Pushed a commit that refactors `BoundAwaitExpression` and extracts an `AwaitableInfo`.

---
In reply to: [161033778](https://github.com/dotnet/roslyn/pull/23961#discussion_r161033778) [](ancestors = 161033778,160565220)